node/libs/roles/src/attester/tests.rs

```diff
@@ -0,0 +1,234 @@
+    let rng = &mut ctx.rng();
+    test_encode_random::<SignedBatchMsg<L1Batch>>(rng);
+    test_encode_random::<L1BatchQC>(rng);
+    test_encode_random::<Msg>(rng);
+    test_encode_random::<MsgHash>(rng);
+    test_encode_random::<Signers>(rng);
+    test_encode_random::<PublicKey>(rng);
+    test_encode_random::<Signature>(rng);
+    test_encode_random::<AggregateSignature>(rng);
+}
+
+#[test]
+fn test_signature_verify() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    let msg1: MsgHash = rng.gen();
+    let msg2: MsgHash = rng.gen();
+
+    let key1: SecretKey = rng.gen();
+    let key2: SecretKey = rng.gen();
+
+    let sig1 = key1.sign_hash(&msg1);
+
+    // Matching key and message.
+    sig1.verify_hash(&msg1, &key1.public()).unwrap();
+
+    // Mismatching message.
+    assert!(sig1.verify_hash(&msg2, &key1.public()).is_err());
+
+    // Mismatching key.
+    assert!(sig1.verify_hash(&msg1, &key2.public()).is_err());
+}
+
+#[test]
+fn test_agg_signature_verify() {
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    let msg1: MsgHash = rng.gen();
+    let msg2: MsgHash = rng.gen();
+
+    let key1: SecretKey = rng.gen();
+    let key2: SecretKey = rng.gen();
+
+    let sig1 = key1.sign_hash(&msg1);
+    let sig2 = key2.sign_hash(&msg2);
+
+    let agg_sig = AggregateSignature::aggregate(vec![&sig1, &sig2]);
+
+    // Matching key and message.
+    agg_sig
+        .verify_hash([(msg1, &key1.public()), (msg2, &key2.public())].into_iter())
+        .unwrap();
+
+    // Mismatching message.
+    assert!(agg_sig
+        .verify_hash([(msg2, &key1.public()), (msg1, &key2.public())].into_iter())
+        .is_err());
+
+    // Mismatching key.
+    assert!(agg_sig
+        .verify_hash([(msg1, &key2.public()), (msg2, &key1.public())].into_iter())
+        .is_err());
+}
+
+fn make_l1_batch_msg(rng: &mut impl Rng) -> L1Batch {
+    L1Batch {
+        number: BatchNumber(rng.gen()),
+        timestamp: time::UNIX_EPOCH + time::Duration::seconds(rng.gen_range(0..1000000000)),
+    }
+}
+
+#[test]
+fn test_l1_batch_qc() {
+    use L1BatchQCVerifyError as Error;
+    let ctx = ctx::test_root(&ctx::RealClock);
+    let rng = &mut ctx.rng();
+
+    let setup1 = Setup::new(rng, 6);
+    let setup2 = Setup::new(rng, 6);
+    let mut genesis3 = (*setup1.genesis).clone();
+    genesis3.attesters_committee =
+        attester::Committee::new(setup1.genesis.attesters_committee.iter().take(3).cloned())
+            .unwrap();
+    let genesis3 = genesis3.with_hash();
+    let attester_weight = setup1.genesis.attesters_committee.total_weight() / 6;
+
+    for i in 0..setup1.attester_keys.len() + 1 {
+        let mut qc = L1BatchQC::new(make_l1_batch_msg(rng), &setup1.genesis);
+        for key in &setup1.attester_keys[0..i] {
+            qc.add(&key.sign_batch_msg(qc.message.clone()), &setup1.genesis);
+        }
+        let expected_weight = i as u64 * attester_weight;
```

Comment:
I asked Esteban to fix this, but I think he forgot: don't unnecessarily assume that all attesters have the same weight here - just sum the weights of first i attesters in the committee.